sel/src/Sel/SecretKey/Stream.hs

```diff
@@ -461,6 +479,47 @@ messageTagToConstant = \case
   Push -> fromIntegral cryptoSecretStreamXChaCha20Poly1305TagPush
   Rekey -> fromIntegral cryptoSecretStreamXChaCha20Poly1305TagRekey
 
+-- | Additional data (AD). Also known as \"additional authenticated data\"
+-- (AAD).
+--
+-- This refers to non-confidential data which is authenticated along with the
+-- message, but not encrypted (i.e. its integrity is protected, but it is not
+-- made confidential).
+--
+-- A typical use case for additional data is to authenticate protocol-specific
+-- metadata about a message, such as its length and encoding.
+newtype AdditionalData = AdditionalData StrictByteString
+  deriving stock (Show, Eq)
+  deriving (Display) via (ShowInstance AdditionalData)
+
+-- | Convert an 'AdditionalData' value to hexadecimal-encoded 'Text'.
+additionalDataToHexText :: AdditionalData -> Base16 Text
+additionalDataToHexText = Base16.encodeBase16 . additionalDataToBinary
+
+-- | Convert an 'AdditionalData' value to a hexadecimal-encoded
+-- 'StrictByteString'.
+additionalDataToHexByteString :: AdditionalData -> Base16 StrictByteString
+additionalDataToHexByteString = Base16.encodeBase16' . additionalDataToBinary
+
+-- | Convert an 'AdditionalData' value to a raw binary 'StrictByteString'.
+additionalDataToBinary :: AdditionalData -> StrictByteString
+additionalDataToBinary (AdditionalData bs) = bs
+
+-- | Error decoding 'AdditionalData' from hexadecimal-encoded bytes.
+newtype AdditionalDataHexDecodingError = AdditionalDataHexDecodingError Text
+  deriving stock (Show, Eq)
+
+-- | Construct an 'AdditionalData' value from a hexadecimal-encoded
+-- 'StrictByteString' that you have obtained on your own, usually from the
+-- network or disk.
+additionalDataFromHexByteString
+  :: Base16 StrictByteString
+  -> Either AdditionalDataHexDecodingError AdditionalData
+additionalDataFromHexByteString hexBs =
+  case Base16.decodeBase16Untyped (Base16.extractBase16 hexBs) of
```

Comment:
This pattern of calling `decodeBase16Untyped . extractBase16` instead of `decodeBase16` on a `Base16 a` value confuses me a bit (https://github.com/haskell-cryptography/libsodium-bindings/pull/144#discussion_r1962461202), but I've done this here to be consistent with similar functions in this module.